Adjust error-handling for invalid filter primitive references

Ahmad-S792 opened the pull request:
#### 8e028375b9c5a9717932ec07c446b4d1b9704345
<pre>
Adjust error-handling for invalid filter primitive references

<a href="https://bugs.webkit.org/show_bug.cgi?id=250276">https://bugs.webkit.org/show_bug.cgi?id=250276</a>
<a href="https://rdar.apple.com/problem/104262208">rdar://problem/104262208</a>

Reviewed by Simon Fraser and Darin Adler.

This patch aligns LegacySVG engine in WebKit with
Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/ffa39ae78904542a2864e2cb7d1c16c24ee2324e">https://chromium.googlesource.com/chromium/blink/+/ffa39ae78904542a2864e2cb7d1c16c24ee2324e</a>

When a filter primitive has an &apos;in&apos; (or, where applicable, &apos;in2&apos;) attribute
that references a non-existent result, treat that reference as having the
default value (==no value) instead of producing an error.

* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::getNamedNode const):
* LayoutTests/svg/filters/in-attribute-error-handling.html: Add Test Case
* LayoutTests/svg/filters/in-attribute-error-handling-expected.html: Add Test Case Expectation
* LayoutTests/css3/filters/invalid-reference-filter-in-chain.html: Add Pixel Tolerance for GTK and WPE

&gt; Rebaselines:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:

Canonical link: <a href="https://commits.webkit.org/279421@main">https://commits.webkit.org/279421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3daa0c92a5f8f6efeb78874aa5f7adae7990b307

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32821 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56745 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4191 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3970 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2740 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31005 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46197 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24470 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3517 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2347 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58340 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28622 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3686 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29829 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46393 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11648 "Found 1 new test failure: fast/dom/intersection-observer-document-leak.html (failure)") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7864 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29599 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
```diff
@@ -77,17 +78,18 @@ class SVGFilterGraph {
-            if (m_lastNode)
-                return m_lastNode;
+        if (!id.isEmpty()) {
+            if (m_sourceNodes.contains(id))
```
I know you just moved it, but this is not great; there are two hash lookups here when there could be just one, if the code used `find()` or `getOptional()`.

Ahmad-S792 replied:
@smfr - I addressed your feedback and opted to use `find()`, if you can have quick scan, if OK - I will directly add 'merge-queue'.

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
```diff
@@ -77,17 +78,21 @@ class SVGFilterGraph {
+        if (!id.isEmpty()) {
+            auto sourceNodes = m_sourceNodes.find(id);
+            if (sourceNodes != m_sourceNodes.end())
+                return m_sourceNodes.get(id);
```
`HashMap::find()` returns an iterator. So we usually name the return value `it` or `iter` or `iterator`. Also `find()` is a look-up and `get()` is another lookup. You can use something like this:

```
auto it = m_sourceNodes.find(id);
if (it != m_sourceNodes.end())
    return it->value.ptr();
```

Or simpler like this:

```
if (auto node = m_sourceNodes.get(id))
    return node;
```

Ahmad-S792 replied:
I will compile locally simpler version and confirm, it progresses and then push another update. :-)

Ahmad-S792 replied:
@shallawa - pushed new build with simpler version. :-)

Thanks for your help as always.

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
```diff
@@ -77,17 +78,19 @@ class SVGFilterGraph {
-            if (m_lastNode)
-                return m_lastNode;
+        if (!id.isEmpty()) {
+            if (auto sourceNodes = m_sourceNodes.get(id))
```
`m_sourceNodes.get()` does not return `sourceNodes` it returns just one `node`.

darinadler replied:
So this local variable should be named `sourceNode` or `node`.

Ahmad-S792 replied:
Opted for `sourceNode` and pushed newer build. :-)

darinadler commented on Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
```diff
@@ -77,17 +78,19 @@ class SVGFilterGraph {
+                return sourceNodes;
 
-            return sourceGraphic();
+            if (auto namedNodes = m_namedNodes.get(id))
```
And this variable should be named `namedNode` or `node`.

Ahmad-S792 replied:
Opted for `sourceNode` and `namedNode` and pushed newer commit. :-)

Commit:
Adjust error-handling for invalid filter primitive references

https://bugs.webkit.org/show_bug.cgi?id=250276
rdar://problem/104262208

Reviewed by Simon Fraser and Darin Adler.

This patch aligns LegacySVG engine in WebKit with
Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/blink/+/ffa39ae78904542a2864e2cb7d1c16c24ee2324e

When a filter primitive has an 'in' (or, where applicable, 'in2') attribute
that references a non-existent result, treat that reference as having the
default value (==no value) instead of producing an error.

* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::getNamedNode const):
* LayoutTests/svg/filters/in-attribute-error-handling.html: Add Test Case
* LayoutTests/svg/filters/in-attribute-error-handling-expected.html: Add Test Case Expectation
* LayoutTests/css3/filters/invalid-reference-filter-in-chain.html: Add Pixel Tolerance for GTK and WPE

> Rebaselines:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:
* LayoutTests/svg/filters/feMerge-wrong-input-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-background-01-f-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-composite-03-f-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-01-b-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-02-b-manual-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/import/filters-overview-03-b-manual-expected.txt:

Canonical link: https://commits.webkit.org/279421@main
